### PR TITLE
Create framework for apps and add constant time stepping

### PR DESCRIPTION
**Pull request by lxmota**
Re-structured things in preparation for adding more patch tests and one or two exemplars.

Reworked main time loop and split into two:
1) A time loop for strictly constant time stepping, for use mainly in debugging.
2) A variable time step loop. I changed the way the output is done. Instead of forcing the output at specific times, the output is now performed as soon as the requested output time is exceeded, without changing the time step. Although forcing the output at specific times is very neat and useful for creating movies, it introduces numerical problems because often the time step is changed to a value many orders of magnitude smaller to what it previously was. This is very undesirable for a multi-step integrator like the OTM one.

Also, I believe there might be a problem in the computation of the stable time step, as the following sequence of steps show:

[hmin.log](https://github.com/SNLComputation/lgrtk/files/4501107/hmin.log)

Looking at the lines where HMIN, C, and DT are calculated, it is seen that the minimum distance HMIN is around 0.125 for the first three steps, and then in the fourth step jumps to 0.354.

Also, when using a variable time step, the patch test does nor pass because the deformations are not the expected ones. The relative error in stress jumps from O(1e-12) to O(1e-4). Not sure if this is due to the issue of the variable time step, or something else.

In addition, I had to revert the order of the loops in the computation of the nodal positions. In the latest version that ran in the GPU, the outer loop was on boundaries and the inner loop was on nodes. This led to incorrect velocities at the boundaries. I switched the order to what it was: outer loop nodes and inner loop boundaries. I know that this might not run in the GPU, but I thought I would change it so that I could add the patch test. I need to get my own GPU build going to avoid this situation.

### CODE REVIEW

File: v3/otm_meshless.cpp [hunk 1]
```diff
@@ -356,30 +358,30 @@ void otm_update_nodal_position(state& s) {
   auto const boundary_indices = s.boundaries;
   auto const boundary_to_prescribed_v = s.prescribed_v.cbegin();
   auto const boundary_to_prescribed_dof = s.prescribed_dof.cbegin();
-  for (auto boundary_index : boundary_indices) {
```
Comment (japlews) [inline v3/otm_meshless.cpp:359]: I'm confused about how switching the loop structure causes errors in the boundary conditions. It seems to me that this loop is fully invertible.

Another option would be to keep the boundary conditions on the device, but you would need to hpc::copy them from host to device at initialization.

Comment (japlews) [reply v3/otm_meshless.cpp:359]: @lxmota see my boundary condition PR. This issue should be fixed (please test) and working with CUDA build.
